x/stake/endblock.go

```diff
@@ -252,3 +265,63 @@ func handleMatureUnbondingDelegations(k keeper.Keeper, ctx sdk.Context) ([]types
 
 	return completed, events
 }
+
+const (
+	maxProcessedRefundCount = 50
+)
+
+func handleRefundCrossStake(ctx sdk.Context, k keeper.Keeper) sdk.Events {
+	iterator := k.IteratorAllDelegations(ctx)
+	defer iterator.Close()
+	var refundEvents sdk.Events
+	count := 0
+	for ; iterator.Valid(); iterator.Next() {
+		delegation := types.MustUnmarshalDelegation(k.CDC(), iterator.Key(), iterator.Value())
+		if delegation.CrossStake {
+			result := handleMsgSideChainUndelegate(ctx, types.MsgSideChainUndelegate{
+				DelegatorAddr: delegation.DelegatorAddr,
+				ValidatorAddr: delegation.ValidatorAddr,
+				Amount:        sdk.NewCoin(k.BondDenom(ctx), delegation.GetShares().RawInt()),
+				SideChainId:   k.ScKeeper.BscSideChainId(ctx),
+			}, k)
+			refundEvents = refundEvents.AppendEvents(result.Events)
+
+			count++
+		}
+		if count >= maxProcessedRefundCount {
+			break
+		}
+	}
+	return refundEvents
+}
+
+func closeSideChainChannels(ctx sdk.Context, k keeper.Keeper) sdk.Events {
+	var events sdk.Events
+	if sdk.IsUpgrade(sdk.BCFusionSecondHardFork) {
+		sideChainId := k.ScKeeper.BscSideChainId(ctx)
+		vp := k.GetSideChainTotalVotingPower(ctx, sideChainId)
+		if vp.LTE(sdk.NewDecFromInt(5_000_000)) {
```

Comment:
I think it is not a good condition to disconnect the channels, because it is quite uncontrollable.  We need announce a exact time that we will disable the cross chain communication. 

So I am thinking this maybe a third hardfork?

Reply:
umm, I think make it in `ThirdHardFork` would be better?

Reply:
`ThirdHardFork ` is a bit heavy.  I think it can be  one month just after the `SecondHardfork`